### PR TITLE
upgrade: Add a precheck for cinder-volume backends

### PR DESCRIPTION
**Pull request by jsuchome**
cinder-volume needs to be configured with HA to enable non-disruptive upgrade

### CODE REVIEW

**Comment by jsuchome** [comment]
[(2) Adapt upgrade pre-checks to ensure cinder-volume is HA](https://trello.com/c/I0uNF2Gb/2084-2-adapt-upgrade-pre-checks-to-ensure-cinder-volume-is-ha)

File: crowbar_framework/app/models/api/crowbar.rb [hunk 1]
```diff
@@ -201,15 +201,18 @@ def compute_status
...
+        return { ha_not_installed: true } unless addon_installed? "ha"
         founders = NodeObject.find("pacemaker_founder:true AND pacemaker_config_environment:*")
-        founders.empty? ?
-          { errors: [I18n.t("api.upgrade.prechecks.ha_configured.not_configured")] }
-          : {}
+        return { ha_not_configured: true } if founders.empty?
+
+        # Check if cinder-volume is highly available
+        prop = Proposal.where(barclamp: "cinder").first
+        elements = prop["deployment"]["cinder"]["elements"]["cinder-volume"] rescue []
```
Comment (jsuchome) [inline crowbar_framework/app/models/api/crowbar.rb]: I think it makes sense here

**Comment by jsuchome** [comment]
@skazi0 New precheck error label for UI...

File: crowbar_framework/app/models/api/crowbar.rb [hunk 2]
```diff
@@ -201,15 +201,18 @@ def compute_status
...
+        non_clustered = elements.select { |e| !ServiceObject.is_cluster? e }
```
Comment (rhafer) [inline crowbar_framework/app/models/api/crowbar.rb]: nit: I believe you could use ```elements.any? { |e| !ServiceObject.is_cluster? e }``` here which would return a true/false value.

Comment (jsuchome) [inline crowbar_framework/app/models/api/crowbar.rb]: true

Comment (jsuchome) [inline crowbar_framework/app/models/api/crowbar.rb]: Or not, i need the value of `non_clustered` as a return value anyway....

Comment (rhafer) [inline crowbar_framework/app/models/api/crowbar.rb]: Ah, right. Not sure why I overlooked that, sorry.

**Comment by jsuchome** [comment]
As we've discussed recently, having cinder-volume HA might not actually be what we need for non-disruptive upgrade. 

Maybe we just need to check for role allocation of cinder-volume roles?

Or should we do some specific check for backends? Something like: RADOS backend is always good, even without extra HA? Or, raw backend is always bad and so on.

File: crowbar_framework/app/models/api/crowbar.rb [hunk 3]
```diff
@@ -201,15 +201,21 @@ def compute_status
...
+        return { ha_not_configured: true } if founders.empty?
+
+        # Check if cinder is using correct backend enabling live-migration
+        prop = Proposal.where(barclamp: "cinder").first
+
+        backends = prop["attributes"]["cinder"]["volumes"].select do |volume|
+          backend_driver = volume["backend_driver"]
+          ["local", "raw"].include? backend_driver
+        end
+        return { cinder_backend: true } unless backends.empty?
```
Comment (Itxaka) [inline crowbar_framework/app/models/api/crowbar.rb]: minor nit: use a proper var that explains what is checked. cinder_backend = true does not explain what this var represents.

**Comment by Itxaka** [comment]
oh wait, wip label was removed!

**Comment by jsuchome** [comment]
@Itxaka i think I've addressed the issues

**Comment by jsuchome** [comment]
> Good to me once the hound issues are resolved.

I'm reluctant to 'resolve' these. I think it's gonna fail with some other problem when I follow the recommendation and current way is sync with what we already have in those files...

File: crowbar_framework/app/models/api/crowbar.rb [hunk 4]
```diff
@@ -220,15 +220,21 @@ def compute_status
...
+        return { ha_not_configured: true } if founders.empty?
+
+        # Check if cinder is using correct backend enabling live-migration
+        prop = Proposal.where(barclamp: "cinder").first
+
+        backends = prop["attributes"]["cinder"]["volumes"].select do |volume|
+          backend_driver = volume["backend_driver"]
+          ["local", "raw"].include? backend_driver
+        end
+        return { cinder_wrong_backend: true } unless backends.empty?
```
Comment (Itxaka) [inline crowbar_framework/app/models/api/crowbar.rb]: :+1:

**Pull request merged**